src/java/org/apache/cassandra/cql3/StatementSource.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.cql3;
+
+import java.util.Objects;
+import java.util.regex.Pattern;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import com.google.common.base.Preconditions;
+
+import org.antlr.runtime.Token;
+import org.apache.cassandra.cql3.statements.QualifiedStatement;
+
+public class StatementSource
+{
+    public static final StatementSource INTERNAL = Bounds.INTERNAL.describeSource(null);
+
+    public final String text;
+    public final Bounds bounds;
+
+    public StatementSource(@Nullable String text, @Nonnull Bounds bounds)
+    {
+        Objects.requireNonNull(bounds);
+        this.text = text;
+        this.bounds = bounds;
+    }
+
+    public String describe(boolean mayIncludeSourceText)
+    {
+        if (this == INTERNAL)
+        {
+            return "<<<internal statement>>>";
+        }
+        else
+        {
+            if (bounds != Bounds.INTERNAL && !bounds.isEmpty() && text != null && mayIncludeSourceText)
+                return String.format("at %d:%d - %s", bounds.startLine + 1, bounds.startChar + 1, text);
+            else if (bounds != Bounds.INTERNAL && !bounds.isEmpty())
+                return String.format("at %d:%d", bounds.startLine + 1, bounds.startChar + 1);
+            else if (text != null && !bounds.isEmpty() && mayIncludeSourceText)
+                return String.format("%s", text);
+            else
+                return "";
+        }
+    }
+
+    public static Bounds createBounds(Token startToken, Token endToken)
+    {
+        Objects.requireNonNull(startToken);
+        Objects.requireNonNull(endToken);
+
+        int startLine, startChar, endLine, endChar;
+
+        if (startToken.getType() == Token.EOF)
+        {
+            startLine = Character.MAX_VALUE;
```

Comment:
if start token is EOF do we not know that endToken is as well?

Reply:
Yes, we can assume that